File: Dynamic_Programming/python/11_가장큰증가부분수열.py
# fast campus 강의
# https://www.acmicpc.net/problem/11055
# 2

from copy import deepcopy

N = int(input())
arr = list(map(int, input().split()))
dp = deepcopy(arr)

for i in range(1, N):
    for j in range(i):
        if arr[j] < arr[i]:
            dp[i] = max(dp[i], dp[j] + arr[i])
print(max(dp))

File: Dynamic_Programming/c++/11_가장큰증가부분수열.cpp
// fast campus 강의
// https://www.acmicpc.net/problem/11055
// 2

#include <iostream>
#include <algorithm>

using namespace std;

int N;
int dp[1001];
int arr[1001];

int main() {
	cin >> N;
	for (int i = 0; i < N; i++) {
		cin >> arr[i];
		dp[i] = arr[i];
	}

	for (int i = 0; i < N; i++) {
		for (int j = 0; j < i; j++) {
			if (arr[j] < arr[i])
				dp[i] = max(dp[i], dp[j] + arr[i]);
		}
	}

	cout << *max_element(dp, dp + N);

}